ui/cypress/e2e/queryBuilder.test.ts

```diff
@@ -0,0 +1,140 @@
+import {Organization} from '../../src/types'
+
+const generateRandomSixDigitNumber = () => {
+  const digits = []
+  for (let i = 0; i < 6; i++) {
+    digits[i] = Math.floor(Math.random() * Math.floor(10))
+  }
+  return Number(digits.join(''))
+}
+
+describe('The Query Builder', () => {
+  beforeEach(() => {
+    cy.flush()
+
+    cy.signin().then(({body}) => {
+      cy.wrap(body.org).as('org')
+    })
+
+    cy.writeData([
+      `mem,host=thrillbo-swaggins active=${generateRandomSixDigitNumber()}`,
+      `mem,host=thrillbo-swaggins cached=${generateRandomSixDigitNumber()}`,
+
+      `mem,host=thrillbo-swaggins active=${generateRandomSixDigitNumber()}`,
+      `mem,host=thrillbo-swaggins cached=${generateRandomSixDigitNumber()}`,
+    ])
+  })
+
+  describe('from the data explorer screen', () => {
+    it('creates a query, edits it to add another field, then views its results with pride and satisfaction', () => {
+      cy.get('@org').then((org: Organization) => {
+        cy.visit(`orgs/${org.id}/data-explorer`)
+      })
+
+      cy.contains('mem').click('right') // users sometimes click in random spots
+      cy.contains('active').click('bottomLeft')
+
+      cy.getByTestID('empty-graph--no-queries').should('exist')
+
+      cy.contains('Submit').click()
+
+      cy.get('.giraffe-plot').should('exist')
+
+      cy.contains('Save As').click()
+
+      // open the dashboard selector dropdown
+      cy.contains('Choose at least 1 dashboard').click()
+      cy.contains('Create a New Dashboard').trigger('mousedown')
+      cy.getByTestID('overlay--container').click() // close the blast door i mean dropdown
+
+      cy.get('[placeholder="Add dashboard name"]').type('Basic Ole Dashboard')
+      cy.get('[placeholder="Add optional cell name"]').type('Memory Usage')
+      cy.contains('Save as Dashboard Cell').click()
+
+      // wait for the notification since it's highly animated
+      // we close the notification since it contains the name of the dashboard and interfers with cy.contains
+      cy.wait(250)
+      cy.get('.notification-close').click()
+      cy.wait(250)
+
+      // force a click on the hidden dashboard nav item (cypress can't do the hover)
+      // i assure you i spent a nonzero amount of time trying to do this the way a user would
+      cy.getByTestID('nav-menu_dashboard').click({force: true})
+
+      cy.contains('Basic Ole Dashboard').click()
```

Comment:
this is a good use of contains. think you can pull it out into a `const` though? to ENSURE it's the same across definition and use?

Reply:
@drdelambre 
i'd prefer to use a string rather than a variable - even though it creates duplication and the there is the chance of making an error, i think it makes the test much more readable, and i often find myself not wanting to put the context of a variable in my head when reading a test, i'd rather just understand exactly and stupidly what's going on by reading it line by line.

if you're super firm on a variable, i can change it though - it's not a crazy strong preference of mine.

Reply:
Not super firm in this instance, but I'll get more opinionated as the number of instances grows in a test